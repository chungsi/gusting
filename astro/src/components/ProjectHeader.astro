---
import CategoryIcon from './CategoryIcon.astro'

import type { CollectionEntry } from 'astro:content'

// Pass the project data as props?
type Props = {
  project: CollectionEntry<'projects'>
  class?: string
}

const { project, class: className }: Props = Astro.props
const { title, category, subtitle, tags } = project.data
---

<header class:list={['relative', className]}>
  <CategoryIcon
    category={category}
    class='absolute w-7 -top-8 left-2 lg:w-8 lg:top-2 lg:left-[-4.5rem]'
  />
  <h1 class='relative inline-block mb-xs'>
    {title}
    <span class='font-mono italic font-light text-sm inline-block p-[0_1em]'>
      &#123;{category}&#125;
    </span>
    <!-- Decorative highlight -->
    <span
      transition:name={`project-highlight-${project.id}`}
      aria-hidden
      class:list={[
        'absolute block top-[-12px] left-[-10px] w-full h-full',
        'rotate-[-2deg] z-[-1] bg-th-highlight',
      ]}
    ></span>
  </h1>
  <p class='italic text-lg tracking-wider'>{subtitle}</p>
  <p class='text-sm'>
    {tags.map((tag) => `${tag} / `)}
  </p>
  <!-- TODO: Also a header image goes here? -->
</header>
