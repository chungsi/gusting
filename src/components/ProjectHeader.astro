---
import CategoryIcon from './CategoryIcon.astro'

// Pass the project data as props?
type Props = {
  title: string
  category: string
  subtitle: string
  tags: string[]
  class?: string
}

const { title, category, subtitle, tags, class: className }: Props = Astro.props
---

<header class:list={['relative', className]}>
  <CategoryIcon
    category={category}
    class='absolute -top-8 left-2 w-7 lg:top-2 lg:left-[-4.5rem] lg:w-8'
  />
  <h1 class='mb-xs relative inline-block'>
    {title}
    <span class='font-th-mono inline-block p-[0_1em] text-sm font-light italic'>
      &#123;{category}&#125;
    </span>
    <!-- Decorative highlight -->
    <!-- transition:name={`project-highlight-${project.id}`} -->
    <span
      aria-hidden
      class:list={[
        'absolute top-[-12px] left-[-10px] block h-full w-full',
        'bg-th-highlight z-[-1] rotate-[-2deg]',
      ]}
    ></span>
  </h1>
  <p class='text-lg tracking-wider italic'>{subtitle}</p>
  <p class='text-sm'>
    {tags.map((tag) => `${tag} / `)}
  </p>
  <!-- TODO: Also a header image goes here? -->
</header>
