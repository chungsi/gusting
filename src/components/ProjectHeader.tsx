import * as React from 'react'
import { getSrc, IGatsbyImageData } from 'gatsby-plugin-image'

import CategoryIcon from './CategoryIcon'

import { concat } from '../utils/helpers'

type ProjectHeaderProps = {
  frontmatter: {
    title: string | null
    subtitle?: string | null
    category?: string | null
    tags?: readonly (string | null)[] | null
    heroImage?: object | null
  }
  withHeroImage?: boolean
  className?: string
}

const ProjectHeader = ({
  frontmatter: { title, subtitle, category, tags, heroImage },
  withHeroImage,
  className
}: ProjectHeaderProps) => {

  // var subtitleFormatClasses = concat(
  //   'pr-10 md:pl-0'
  // )

  return (
    <header className={concat('relative', className ?? '')}>
      {category &&
        <CategoryIcon
          category={category}
          className={concat(
            'absolute w-7 -top-8 left-2',
            'lg:w-8 lg:top-2 lg:left-[-4.5rem]'
          )}
        />
      }

      <h1 className='relative inline-block mb-xs'>
        {title}

        {category &&
          <span className={concat(
            'font-mono italic font-light text-sm',
            'inline-block p-[0_1em]'
          )}>
            &#123;{category}&#125;
          </span>
        }

        <span
          aria-hidden
          className={concat(
            'absolute block top-[-12px] left-[-10px] w-full h-full',
            'rotate-[-2deg] z-[-1] bg-th-highlight'
          )}
        />
      </h1>

      {subtitle &&
        <p className={concat('italic text-lg')}>{subtitle}</p>
      }

      {tags &&
        <p className={concat('text-sm')}>
          {tags.map(tag => `${tag} / `)}
        </p>
      }

      {withHeroImage && heroImage &&
        <img
          aria-hidden
          src={getSrc(heroImage as IGatsbyImageData)}
          alt=''
          className={concat(
            'absolute right-[-2rem] top-[-1rem] z-[-1] opacity-70',
            'block max-w-[7rem]',
            'md:max-w-[10rem]',
          )}
        />
      }
    </header>
  )
}

export default ProjectHeader