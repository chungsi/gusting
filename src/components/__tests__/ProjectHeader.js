import React from 'react'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'

import ProjectHeader from '../ProjectHeader'

// Using react's testing thing
describe('ProjectHeader', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<ProjectHeader frontmatter={{
        title: 'Project Header',
        subtitle: 'test project header',
        category: 'art',
        tags: ['one', 'two', 'three'],
      }} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

// Using the other library for testing components
// const TestProjectHeader = () => <ProjectHeader
//   frontmatter={{
//     title: 'Project Header',
//     subtitle: 'test project header',
//     category: 'art',
//     tags: ['one', 'two', 'three'],
//   }}
//   data-testid="project-header"
// />

// test("Display correct project header", () => {
//   const { getByTestId } = render(<TestProjectHeader />)
//   // Assertion
//   expect(getByTestId('project-header')).toHaveTextContent('Project Header')
// })