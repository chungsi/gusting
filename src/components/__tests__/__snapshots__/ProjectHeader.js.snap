// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectHeader renders correctly 1`] = `
<header
  className=""
>
  <div
    aria-hidden={true}
    className="absolute w-7 -top-8 left-2 lg:w-8 lg:top-2 lg:left-[-4.5rem]"
  >
    <test-file-stub />
  </div>
  <h1
    className="relative inline-block mb-xs"
  >
    Project Header
    <span
      className="font-mono italic font-light text-sm inline-block p-[0_1em]"
    >
      {
      art
      }
    </span>
    <span
      aria-hidden={true}
      className="absolute block top-[-12px] left-[-10px] w-full h-full rotate-[-2deg] z-[-1] bg-th-highlight"
    />
  </h1>
  <p
    className="italic text-lg"
  >
    test project header
  </p>
  <p
    className="text-sm"
  >
    one / 
    two / 
    three / 
  </p>
</header>
`;
